### ISSUE TITLE
torch.sum(tensor, dim=()) is different from np.sum(arr, axis=())

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#29137]
## 🐛 Bug

`torch.sum(tensor, dim=())` performs a full reduce, while `np.sum(arr, axis=())` performs no reduce.

## To Reproduce

```
import torch
import numpy as np
arr = np.array([1, 2, 3])
tensor = torch.from_numpy(arr)

tensor.sum(())  # gives tensor(6)
arr.sum(())  # gives np.array([1, 2, 3])
```
## Expected behavior

These should probably be the same. Also, I don't think it makes sense for `tensor.sum(dim=())` to do a full reduce. This special casing caused the bug at https://github.com/pytorch/pytorch/issues/28993 and I suspect the gradient for things like torch.var may be incorrect right now.

## Environment

Pytorch master

## Additional context

Some people who have been involved in multidim support who may be interested: @t-vi, @SsnL, @umanwizard @gchanan. Please let me know your opinions on this subject.

cc @ezyang @gchanan @zou3519 @jerryzh168 @SsnL

**Comment by t-vi** [comment]
So this seems to have been introduced with the switch to TensorIterator (#11908) and may be an inadvertent result of the TensorIterator logic of specifying dimensions.
Before, passing `[]` would mean no reduction.
https://github.com/pytorch/pytorch/pull/11908/files#diff-847680695fb0535da86db8b98a13b5a3L523

**Comment by VitalyFedyunin** [comment]
Probably need to remove from triage pipeline.

**Comment by umanwizard** [comment]
Yes this is a known issue. The problem is that downstream functions take an IntArrayRef and don't know how to distinguish between an empty one and one having not been specified at all (e.g. in the `torch.mean(x)` case). Should be fixed IMO by taking `option<IntArrayRef>` everywhere; one issue is what `torch.mean(x, dims=None)` should mean.

I vaguely think I discussed this with @colesbury at some point but I could be wrong.

**Comment by zou3519** [comment]
fwiw in numpy, `np.mean(x, dims=None)` performs a full reduce while `np.mean(x, dims=())` performs no reduce

**Comment by gchanan** [comment]
We need to decide if we should just change the gradient formulas or change our implementations to have the same behavior as numpy (probably would need to deprecation / fix cycle).

Also, do we still have any reductions that support multiple dimensions that don't use TensorIterator?  Because they would have different behavior.

**Comment by ezyang** [comment]
Triage review likes lining up with the numpy behavior

**Comment by kurtamohler** [comment]
@pbelevich, have you started on this issue? If not, I could take over if you'd like. I wanted to work on #28993, but it's blocked by this.

**Comment by pbelevich** [comment]
@kurtamohler I didn't start working on this issue. Since I'm quite busy with other tasks, please take this issue.

**Comment by kurtamohler** [comment]
cool, thanks!

**Comment by kurtamohler** [comment]
I think it makes a lot of sense to use option<IntArrayRef>, as umanwizard suggested. Right now, I'm just focusing on getting torch.sum to work properly, and then I can apply the same fix to the other reduction functions. So far I've gathered that I need to make at least these changes:

- Update the appropriate sum kernels in aten/src/ATen/native/ReduceOps.cpp to use option<IntArrayRef>

- Update aten/src/ATen/native/native_functions.yaml to use "IntArrayRef?" rather than "int[1]" for the sum dim argument

- Add "IntArrayRef?" to the type map in tools/jit/gen_jit_dispatch.py, since it's not already there

- Update autograd to use the optional type as well. This involves updating sum.dim_IntList in tools/autograd/derivatives.yaml to use "IntArrayRef?" instead of "int[1]", and also updating "sum_backward" in tools/autograd/templates/Functions.cpp to use optional<IntArrayRef>.

It took some trial and error to figure out those things, and I now have a build error that I'm having trouble figuring out.

`
[3/595] Building CXX object caffe2/CMakeFiles/torch.../torch/csrc/jit/generated/register_aten_ops_0.cpp.o
FAILED: caffe2/CMakeFiles/torch.dir/__/torch/csrc/jit/generated/register_aten_ops_0.cpp.o 
...
In file included from ../aten/src/ATen/core/ivalue.h:651:0,
                 from ../aten/src/ATen/core/stack.h:3,
                 from ../torch/csrc/jit/operator.h:6,
                 from ../torch/csrc/jit/generated/register_aten_ops_0.cpp:1:
../aten/src/ATen/core/ivalue_inl.h: In instantiation of 'T c10::generic_to(c10::IValue, c10::_fake_type<T>) [with T = c10::ArrayRef<long int>]':
../aten/src/ATen/core/ivalue_inl.h:576:20:   required from 'T c10::IValue::to() const & [with T = c10::ArrayRef<long int>]'
../aten/src/ATen/core/ivalue_inl.h:770:22:   required from 'c10::optional<T> c10::IValue::toOptional() [with T = c10::ArrayRef<long int>]'
../torch/csrc/jit/generated/register_aten_ops_0.cpp:5852:73:   required from here
../aten/src/ATen/core/ivalue_inl.h:494:73: error: no type named 'element_type' in 'std::__remove_pointer_helper<c10::ArrayRef<long int>, c10::ArrayRef<long int> >::type {aka class c10::ArrayRef<long int>}'
     using ElemType = typename std::remove_pointer<T>::type::element_type;
                                                                         ^
`

I think I probably have to add something to either aten/src/ATen/core/ivalue_inl.h or tools/jit/templates/register_aten_ops.cpp, but I'm not yet sure exactly what.

Also, could anyone tell me, at least in general terms, what other steps I'll have to do? I'm discovering everything piece by piece, so it would be very helpful to have a full overview of what all needs to be updated when changing a kernel's argument type.

**Comment by t-vi** [comment]
Not that I know anything about this, but in
> Update aten/src/ATen/native/native_functions.yaml to use "IntArrayRef?" rather than "int[1]" for the sum dim argument

you probably want to handle the case of an int being passed (the `[1]` bit) -- quite possibly these days this can be achieved by introducing an int overload (two years ago this didn't work).

As far as I understand (which is not terribly far), you'd be breaking the ABI by adding the optional, but the same code would probably still compile.
Also, you might need to add support for the type hints.

**Comment by ezyang** [comment]
You may have met an unfortunate fate. For two reasons:

1. I did some grepping in `native_functions.yaml` and I didn't see any occurrences of `int[1]?` (which is the type name I would have expected to desugar to `optional<IntArrayRef>` in C++). So it's entirely possibly that this is not supported at all.
2. Even if it did work, it appears you picked the wrong type name and have spent a bunch of time trying to add support for a type name we don't want. I reviewed your description of changes and unfortunately I don't think any of them are correct.

Let me do a little more research and I will post some more.

**Comment by ezyang** [comment]
So, **if** `int[1]?` works out of the box, here is the series of steps I expect you to need to do:

1. Update the `native_functions.yaml` so instead of `func: sum.dim_IntList(Tensor self, int[1] dim, bool keepdim=False, *, ScalarType? dtype=None) -> Tensor` you have `func: sum.dim_IntList(Tensor self, int[1]? dim, bool keepdim=False, *, ScalarType? dtype=None) -> Tensor`
2. Update the C++ definitions of the functions in ATen/native to take `optional<IntArrayRef>` instead of `IntArrayRef`
3. No other changes necessary

The change should not be BC-breaking because it should always be valid to make a previously mandatory argument optional.

If `int[1]?` doesn't work, the changes are substantially more involved. Sync back to me in that case.

**Comment by ezyang** [comment]
I'd also suggest reading the README at https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native

**Comment by kurtamohler** [comment]
ok, I'll try that, thanks!

**Comment by kurtamohler** [comment]
@ezyang, I tried what you suggested. I found that I also had to update the `sum.dim_IntList` entry in `tools/autograd/derivatives.yaml`.
Unfortunately, this doesn't seem to work. Here's the error I get:
```
[1/1429] Generating ../../torch/csrc/autograd/gener...c/autograd/generated/python_nn_functions_dispatch.h
FAILED: ../torch/csrc/autograd/generated/Functions.cpp ../torch/csrc/jit/generated/register_aten_ops_0.cpp ../torch/csrc/jit/generated/register_aten_ops_1.cpp ../torch/csrc/jit/generated/register_aten_ops_2.cpp ../torch/csrc/autograd/generated/VariableType_0.cpp ../torch/csrc/autograd/generated/VariableType_1.cpp ../torch/csrc/autograd/generated/VariableType_2.cpp ../torch/csrc/autograd/generated/VariableType_3.cpp ../torch/csrc/autograd/generated/VariableType_4.cpp ../torch/csrc/autograd/generated/Functions.h ../torch/csrc/autograd/generated/variable_factories.h ../torch/csrc/autograd/generated/VariableType.h ../torch/csrc/autograd/generated/python_functions.cpp ../torch/csrc/autograd/generated/python_variable_methods.cpp ../torch/csrc/autograd/generated/python_torch_functions.cpp ../torch/csrc/autograd/generated/python_nn_functions.cpp ../torch/csrc/autograd/generated/python_functions.h ../torch/csrc/autograd/generated/python_variable_methods_dispatch.h ../torch/csrc/autograd/generated/python_torch_functions_dispatch.h ../torch/csrc/autograd/generated/python_nn_functions.h ../torch/csrc/autograd/generated/python_nn_functions_dispatch.h 
cd /home/kurtamohler/development/pytorch && /home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/bin/python tools/setup_helpers/generate_code.py --declarations-path /home/kurtamohler/development/pytorch/build/aten/src/ATen/Declarations.yaml --nn-path aten/src
Skipped writing torch/csrc/autograd/generated/python_functions.h
...
Skipped writing torch/csrc/autograd/generated/variable_factories.h
Traceback (most recent call last):
  File "tools/setup_helpers/generate_code.py", line 93, in <module>
    main()
  File "tools/setup_helpers/generate_code.py", line 88, in main
    options.selected_op_list_path,
  File "tools/setup_helpers/generate_code.py", line 57, in generate_code
    selected_op_list_path=selected_op_list_path)
  File "/home/kurtamohler/development/pytorch/tools/jit/gen_jit_dispatch.py", line 465, in gen_jit_dispatch
    shards[x].append(OPERATOR.substitute(signature=signature(decl, decl['should_match_schema']),
  File "/home/kurtamohler/development/pytorch/tools/jit/gen_jit_dispatch.py", line 560, in signature
    args.append(format_arg(a))
  File "/home/kurtamohler/development/pytorch/tools/jit/gen_jit_dispatch.py", line 528, in format_arg
    typ = jit_type_of(arg)
  File "/home/kurtamohler/development/pytorch/tools/jit/gen_jit_dispatch.py", line 88, in jit_type_of
    typ = TYPE_MAP[arg['simple_type']]
KeyError: 'IntArrayRef?'
```

Since I got rid of `IntArrayRef?` and replaced it with `int[1]?`, I was confused why it was showing up again. But it looks like it's coming from generated code. Searching the entire project:
```
 $ grep -R 'IntArrayRef?' .
./torch/csrc/autograd/generated/python_variable_methods.cpp:    "sum(IntArrayRef?[1] dim, bool keepdim=False, *, ScalarType? dtype=None)",
./torch/csrc/autograd/generated/python_torch_functions.cpp:    "sum(Tensor input, IntArrayRef?[1] dim, bool keepdim=False, *, ScalarType? dtype=None, Tensor out=None)",
```

**Comment by zou3519** [comment]
Instead of using `int[1]?`, couldn't we reuse the existing schemas?
We have int[1] and no dim argument overloads:
```
- func: mean.dim(Tensor self, int[1] dim, bool keepdim=False, *, ScalarType? dtype=None) -> Tensor
- func: mean(Tensor self, *, ScalarType? dtype=None) -> Tensor
```

In the future I think it would be nice to consolidate these schemas to one single `int[1]?` schema but I don't think it is required for the fix

**Comment by ezyang** [comment]
@zou3519 One downside of doing it this way is that `mean(self, dim=None)` doesn't work (whereas, according to the conversation above, it works on Numpy). (It may still be profitable to half fix it in the way you suggested.)

**Comment by kurtamohler** [comment]
I could do a partial fix with the existing schemas--just do a full reduce if the kernel without `dim` is called, and don't reduce at all if the kernel with `dim` is called but `dim` is empty.

However, an additional downside would be that if someone were to run `tensor.sum(keepdim=True)`, dim would get set to an empty tuple, since the kernel that has the `keepdim` argument also has `dim`. In that case, there would be no reduction, whereas we would have wanted a full reduction.

I don't think this is acceptable behavior, and I don't think we could get around it without enabling `int[1]?`. @ezyang, what would be the steps to add support for `int[1]?` ?

**Comment by zou3519** [comment]
> I could do a partial fix with the existing schemas--just do a full reduce if the kernel without `dim` is called, and don't reduce at all if the kernel with `dim` is called but `dim` is empty.
> 
> However, an additional downside would be that if someone were to run `tensor.sum(keepdim=True)`, dim would get set to an empty tuple, since the kernel that has the `keepdim` argument also has `dim`. In that case, there would be no reduction, whereas we would have wanted a full reduction.

Assuming we kept the current schemas, `tensor.sum(keepdim=True)` would error out gracefully because it wouldn't match either of the schemas we have:

```
func: sum(Tensor self, *, ScalarType? dtype=None) -> Tensor
func: sum.dim_IntList(Tensor self, int[1] dim, bool keepdim=False, *, ScalarType? dtype=None) -> Tensor
```
so I think that isn't a concern.

**Comment by kurtamohler** [comment]
@zou3519 , you're right, I just tried running tensor.sum(keepdim=True) and it gave me an error. Alright, I'll go ahead and fix the behavior of the current schemas.

**Comment by kurtamohler** [comment]
Just a head's up, I was able to complete the fix for `sum`, but when I ran the test suite, it became apparent that now some of the other reduction functions are broken, like `prod`.

Much of the logic I added for this fix is in `make_reduction()` in `aten/src/ATen/native/ReduceOps.cpp`, and other reduction kernels call this function. When `dim` is not one of the kernel's arguments, `make_reduction()` is called with dim set to `{}`. I need to change each of these cases to set dim to `c10::nullopt`. Shouldn't be too much trouble to make these changes though, I just have to find all of them.

**Comment by zou3519** [comment]
BTW, before we fix this, we should probably deprecate it first so that users don't suddenly get surprised by the new (correct) behavior. I'm not sure if it is too late to get the deprecation into pytorch 1.4.

**Comment by ezyang** [comment]
@kurtamohler Nice! Do you mind putting up the diff earlier rather than later so we can take a look?

**Comment by kurtamohler** [comment]
So far, the above PR fixes how sum, prod, mean, and norm handle dim=(). There are a few others I need to fix too, and I need to add tests as well. Please let me know if there's anything that could be done in a better way, this is my first contribution to this project after all.

**Comment by kurtamohler** [comment]
I've fixed all the appropriate reduction functions that were broken by my change. Now all the reductions where `dim=()` is legal (the others give an API error) no longer perform a reduction in that case. I added tests for all the legal cases.

**Comment by rgommers** [comment]
Also confirmed that this was a regression between v0.4.1 (which behaves the same as this PR) and v1.0

**Comment by rgommers** [comment]
Update: this is currently blocked on adding a new `ReductionDim` class (see https://github.com/pytorch/pytorch/pull/30822#issuecomment-572783469), to be able to support `dim=None`.

**Comment by teichert** [comment]
@zou3519 Should the ability to handle `dim=None` for reduction along all dimensions (as is done in numpy) be added to the description of this issue or should a new issue be opened for that?

**Comment by kurtamohler** [comment]
I can pick this back up now since PR #43262 added support for `int[n]?`

**Comment by kurtamohler** [comment]
Since #70864 was merged, it should be relatively straightforward to add support for `dim=None` (no reduce), so I will go ahead and do that now.

As for `dim=()`, should we go ahead and deprecate it so we can fix it in the future?

**Comment by rgommers** [comment]
> As for `dim=()`, should we go ahead and deprecate it so we can fix it in the future?

I'd suggest to not do that straight away. When something is deprecated, it's nice to move over to something else straight away (`dim=None` in this case) - and that's a lot harder if that new functionality isn't present in the last release. So unless it's urgent (and it's not here imho) I always prefer to wait 1-2 releases before adding the deprecation.

The important part of this issue is adding support for `dim=None`. I'm not sure we care as much about the semantics of `dim=()` because that's highly non-idiomatic in numpy-using code - it'd be nice to fix, but more gradually.

**Comment by zou3519** [comment]
> I'm not sure we care as much about the semantics of dim=() because that's highly non-idiomatic in numpy-using code 

It's pretty common for PyTorch developers to run into this (at least for me -- I've run into this multiple times since I filed this issue). So we do want to get this fixed eventually but I agree that we shouldn't rush into it

**Comment by kurtamohler** [comment]
Oops, above I said that I would "add support for dim=None (no reduce)". But I was mistaken, `dim=None` is supposed to be an all-reduce, since that's the behavior in NumPy

**Comment by vadimkantorov** [comment]
Is/will this `dim=None` behavior supported for all reductions?

**Comment by kurtamohler** [comment]
Below is a list of the operators whose `dim` args need to be updated to `int[1]? dim` in `aten/src/ATen/native/native_functions.yaml` to enable `dim=None` support. Note that some operators currently have partial `dim=None` support with the schema type `int[1] dim=[]`, but they too need to be changed to `int[1]? dim`. I'll track PRs for each op and their status here.

* `sum` #79881 (merged)
* `mean` #81286 (merged)
* `std` #81845
* `logsumexp`
* `special_logsumexp`
* `amax`
* `nanmean`
* `amin`
* `nansum`
* `std_mean`
* `roll`
* `var`
* `var_mean`
* `_sparse_sum`, `_sparse_csr_sum`, and `_sparse_csr_prod`
* `norm`, `frobenius_norm`, `native_norm`
* `fft_*fft2`

**Comment by zou3519** [comment]
still different, I don't think the team thinks this is hi-pri.
